### ISSUE TITLE
Regression: libsodium-sys 0.0.16 → 0.1.0 fails to build

### ISSUE THREAD

**Issue by zetok** [sodiumoxide/sodiumoxide#245]
```
error: failed to run custom build command for `libsodium-sys v0.1.0`
process didn't exit successfully: `/home/zetok/git/rust/epaste/target/debug/build/libsodium-sys-4fc3d445e3b72076/build-script-build` (exit code: 101)
--- stdout
cargo:rerun-if-env-changed=SODIUM_LIB_DIR
cargo:rerun-if-env-changed=SODIUM_INC_DIR
cargo:rerun-if-env-changed=SODIUM_STATIC
cargo:rustc-link-search=native=/usr/lib64
cargo:rustc-link-lib=sodium

--- stderr
thread 'main' panicked at 'Unable to find libclang: "couldn\'t find any of [\'libclang.so\', \'libclang.so.*\', \'libclang-*.so\'], set the LIBCLANG_PATH environment variable to a path where one of these files can be found (skipped: [])"', libcore/result.rs:945:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

I've got plenty of `libclang.so` on my system:

```
$ find /usr -type f -name 'libclang.so*'
/usr/lib64/llvm/5/lib32/libclang.so.5.0
/usr/lib64/llvm/5/lib64/libclang.so.5.0
/usr/lib64/llvm/6/lib32/libclang.so.6.0
/usr/lib64/llvm/6/lib64/libclang.so.6.0
```

IMO making library consumer set some env variables just to compile library is not a good idea, especially when previously it worked just fine without having to do it.

**Comment by warner** [comment]
Just guessing, but do you maybe have a `libclang` package installed but not a `libclang-dev` package? On some of my systems, it's only the `libclang-dev` package that provides the `.so` symlink to some specific numbered version of the library. (I don't know what platform you're on, `-dev` is the debian/ubuntu suffix, I think RPM-style packages have a different spelling, maybe `-devel`).

Also, I've found that I have to install the actual `clang` package to avoid an unintuitive "cannot find stddef.h" error later in the process. (https://github.com/rust-lang-nursery/rust-bindgen/issues/1270 for details)

I'm with you on the env variables being annoying. 0.0.16 was working on my debian box, and 0.1.0 doesn't, but I think it's because the libsodium that's in debian "stretch" just happens to be compatible with the 0.0.16 bindings, whereas 0.1.0 requires a more recent version.

**Comment by nvesely** [comment]
Can you share the output of `sudo /sbin/ldconfig -p | grep clang`? I'm also thinking this is a linking problem.

I agree this is not an ideal situation, but for debugging purposes can you try setting `export LIBCLANG_PATH=/usr/lib64/llvm/6/lib64/libclang.so.6.0` (or whatever version is most appropriate for your platform)?

**Comment by zetok** [comment]
@warner I'm on Gentoo, and it doesn't strip from packages stuff needed for compiling.

@nvesely 
[ldconfig -p | grep clang.txt](https://github.com/sodiumoxide/sodiumoxide/files/2324071/ldconfig.-p.grep.clang.txt)

`cargo build` still fails after `export LIBCLANG_PATH='/usr/lib64/llvm/6/lib64/libclang.so.6.0'`

**Comment by zetok** [comment]
Actually, exporting just the *path* did the trick and libsodium-sys was built successfully.

Exported path: `export LIBCLANG_PATH='/usr/lib64/llvm/6/lib64/'`

**Comment by nvesely** [comment]
What's the error with sodiumoxide compiling now?

**Comment by zetok** [comment]
[sodiumoxide errors](https://gist.github.com/zetok/1f37eded4d6f3543c4d002d99d97d2dc)

**Comment by zetok** [comment]
libsodium 1.0.11

**Comment by kpcyrd** [comment]
This is probably the same issue as #244. Debian stretch currently ships 1.0.11 and stretch-backports is on 1.0.16.

Gentoo currently lists 1.0.15 as stable, you could try updating to that if possible. 1.0.16 is available as well but would require unmasking.

**Comment by nvesely** [comment]
Short term fix: #248. Long-term fix: #234.

**Comment by nvesely** [comment]
Can you confirm the this builds when using the compiler and library versions specified in the README? @zetok
